### PR TITLE
DEV: Improve MessageBus subscriptions for TopicTrackingState

### PR DESCRIPTION
**Pull request by tgxworld**
## Why do we need this change? 

When loading the ember app, [MessageBus does not start polling immediately](https://github.com/discourse/discourse/blob/f31f0b70f82c43d93220ce6fc0d4f57440452f37/app/assets/javascripts/discourse/app/initializers/message-bus.js#L71-L81) and instead waits for `document.readyState` to be `complete`. What this means is that if there are new messages being created while we have yet to start polling, those messages will not be received by the client.

With sidebar being the default navigation menu, the counts derived from `topic-tracking-state.js` on the client side is prominently displayed on every page. Therefore, we want to ensure that we are not dropping any messages on the channels that `topic-tracking-state.js` subscribes to.  

## What does this change do? 

This includes the `MessageBus.last_id`s for the MessageBus channels which `topic-tracking-state.js` subscribes to as part of the preloaded data when loading a page. The last ids are then used when we subscribe the MessageBus channels so that messages which are published before MessageBus starts polling will not be missed.

## Review Notes

1. See https://github.com/discourse/message_bus#client-support for documentation about subscribing from a given message id.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/models/topic-tracking-state.js [hunk 1]
```diff
@@ -83,20 +83,50 @@ const TopicTrackingState = EmberObject.extend({
    *
    * @method establishChannels
    */
-  establishChannels() {
-    this.messageBus.subscribe("/latest", this._processChannelPayload);
+  establishChannels(meta) {
+    this.messageBus.subscribe(
+      "/latest",
+      this._processChannelPayload,
+      meta ? meta["/latest"] : -1
```
Comment (martin-brennan) [inline app/assets/javascripts/discourse/app/models/topic-tracking-state.js:90]: Kind of nitpicky but unless you have read the MB docs, `-1` is a bit magic, maybe we should put this in a const like `MESSAGE_BUS_ALL_NEW` or something. Just a suggestion, I am mainly concerned about people questioning what this means or changing it to other values which can do different things

File: app/assets/javascripts/discourse/app/models/topic-tracking-state.js [hunk 2]
```diff
@@ -83,20 +83,50 @@ const TopicTrackingState = EmberObject.extend({
    *
    * @method establishChannels
    */
-  establishChannels() {
-    this.messageBus.subscribe("/latest", this._processChannelPayload);
+  establishChannels(meta) {
```
Comment (martin-brennan) [inline app/assets/javascripts/discourse/app/models/topic-tracking-state.js:86]: Small improvement, if you do `meta = {}` here then you don't need to check for meta everywhere below, you can just do:

```
meta["/latest"] || -1
```

Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/models/topic-tracking-state.js:86]: Or if we are worried about `PreloadStore` passing null here, we could just do this as the first thing in the function:

```
meta ??= {}
```

File: app/assets/javascripts/discourse/app/models/topic-tracking-state.js [hunk 3]
```diff
@@ -1005,10 +1035,10 @@ const TopicTrackingState = EmberObject.extend({
 });
 
 export function startTracking(tracking) {
-  const data = PreloadStore.get("topicTrackingStates");
-  tracking.loadStates(data);
-  tracking.establishChannels();
+  tracking.loadStates(PreloadStore.get("topicTrackingStates"));
+  tracking.establishChannels(PreloadStore.get("topicTrackingStateMeta"));
   PreloadStore.remove("topicTrackingStates");
+  PreloadStore.remove("topicTrackingStateMeta");
```
Comment (martin-brennan) [inline app/assets/javascripts/discourse/app/models/topic-tracking-state.js:1041]: For both of these we can just use `PreloadStore.getAndRemove`

File: app/models/topic_tracking_state.rb [hunk 4]
```diff
@@ -31,6 +31,13 @@ class TopicTrackingState
   DISMISS_NEW_MESSAGE_TYPE = "dismiss_new"
   MAX_TOPICS = 5000
 
+  PUBLISH_NEW_MESSAGE_BUS_CHANNEL = "/new"
```
Comment (martin-brennan) [inline app/models/topic_tracking_state.rb:34]: I think `PUBLISH_` on all these is probably redundant

**Review by martin-brennan — APPROVED** [review]
Really nice, the comments I have aren't blockers.

File: app/assets/javascripts/discourse/app/models/topic-tracking-state.js (hunk 1 above)
Comment (tgxworld) [reply app/assets/javascripts/discourse/app/models/topic-tracking-state.js:90]: Actually, I had to do this just for the test enviornment. Let me see if we already have an existing way to preload things in our acceptance test.

File: app/assets/javascripts/discourse/app/models/topic-tracking-state.js (hunk 3 above)
Comment (tgxworld) [reply app/assets/javascripts/discourse/app/models/topic-tracking-state.js:1041]: O nice :+1: I didn't know about this.

**Comment by tgxworld** [comment]
Thank you @martin-brennan for reviewing :+1:

**Pull request merged**